### ISSUE TITLE
Expiring a mostly-finished task run.

### ISSUE THREAD

**Issue by bottler** [facebookresearch/Mephisto#431]
If I have a large task run, expiring it expires all its units, which sets the expiry on mturk even if the unit is completed. This can take a long time. I am wondering if this is worth optimizing. If so, would the following change in mephisto, in `mturk_unit.py`, be safe?

```
     def expire(self) -> float:
         """
         Send a request to expire the HIT, and if it's not assigned return,
         otherwise just return the maximum assignment duration
         """
         delay = 0

+        status = self.get_status()
+        if status in [AssignmentState.EXPIRED, AssignmentState.COMPLETED]:
+            return delay

-        if self.get_status() == AssignmentState.ASSIGNED:
+        if status == AssignmentState.ASSIGNED:
```

**Comment by JackUrb** [comment]
This makes sense to me, completed or expired HITs shouldn't need to be re-expired. The only concern would be ensuring that these units indeed have an associated `HIT` already, but if they have a completed `Agent` associated with them I don't see how you could get here with a `Unit` that has no mapped HIT.

**Comment by bottler** [comment]
I've been testing this out, running large runs for a few days, and it seems to be working. I will make a PR.

There are often cases where I expire a run* and wait a while (much longer than the assignment duration) and find that there's a hit or two in the "launched" state. In this case, expiring the run again succeeds in cleaning up. 

* i.e. `operator.get_running_task_runs()[...].task_launcher.expire_units()`

**Comment by JackUrb** [comment]
Interesting to note that there are hits that are still in LAUNCHED - I wonder if they transitioned _back_ to that state after being expired, or if something else happens for these. I'd imagine that on the MTurk side, the HIT is actually already expired.

**Comment by bottler** [comment]
Yes, they are expired on the mturk side, so something has gone wrong in keeping track of the status inside mephisto. Waiting doesn't lead mephisto to correct itself, though, and I don't think calling `get_status` on them does it either.

**Comment by JackUrb** [comment]
I'm betting that the task gets returned after it has been 'expired', which is causing the transition to the `LAUNCHED` state. Will need to see how to reconcile the transition of a return if the underlying task is no longer available.

**Comment by federicoruggeri** [comment]
I'm not sure if this comment is directly related to above issue. I've noticed that if timeout occurs, then the unit state follows the same state change: from `assigned` to `expired` and then to `launched`. This occurs everytime timeout expires (tested with local and heroku architect and mock agent on dialogue task).

By inspecting the code, i've seen that `Unit.get_status()` function in `unit.py` does not check for `AgentState.STATUS_TIMEOUT`.
If we add the following condition:
`elif agent_status == AgentState.STATUS_TIMEOUT: computed_status = AssignmentState.EXPIRED` (around line 294)

then, we avoid the problem of state overwriting: from `expired` to `launched` again. In particular, the two units in the same assignments will have the following final states: `expired` and `completed`.

I've noticed this due to how the automatic unit assignment works (find the closest unit with `launched` state)  -> btw, is this intended?

Certainly, I'm not sure if this quick-fix is how it should be done.

**Comment by JackUrb** [comment]
@federicoruggeri this is something of a bug in behavior, but not of intent. When an agent disconnects, if the system doesn't explicitly expire the `Unit`, we assume that the `Unit` should be put up for a different agent to be able to accept. In the case of a dialogue task though, we do expire the `Unit` directly, but then the get_status overrides that. We should reject the edge here: though specifically if the unit is already in the expired state:
https://github.com/facebookresearch/Mephisto/blob/fc359c74dc73c4c408b089301833682107bbbbfa/mephisto/data_model/unit.py#L285-L289
